### PR TITLE
Cast policies and documentation update

### PR DESCRIPTION
**Pull request by aTom3333**
This pull request include work done on the cast policies (not over yet) and a big update of the documentation.

### CODE REVIEW

File: .codecov.yml [hunk 1]
```diff
@@ -1,2 +1,7 @@
 ignore:
-  - "example/*"
+  - "cmake-build-debug/**"
+  - "cmake-modules/**"
+  - "doc/**"
+  - "external/**"
+  - "test/**"
```
Comment (sdavtaker) [inline .codecov.yml:6]: Why you ignoring these?

File: .travis.yml [hunk 2]
```diff
@@ -12,35 +12,47 @@ matrix:
...
+        # Uploading report to CodeCov
+        - bash <(curl -s https://codecov.io/bash) || echo "Codecov did not collect coverage reports"
 
   allow_failures:
-          - os: linux
+    - os: osx
 
 script:
-  - cmake -DCMAKE_BUILD_TYPE=Debug -G "Unix Makefiles" . && make && make test  
+  - cmake -DCMAKE_BUILD_TYPE=Debug -DREAL_BENCH=OFF -G "Unix Makefiles" . && make -j`nproc` && make -j`nproc` test
```
Comment (sdavtaker) [inline .travis.yml:58]: What is that REAL_BENCH for?

**Review by sdavtaker — CHANGES_REQUESTED** [review]
2 small comments

File: .codecov.yml (hunk 1 above)
Comment (aTom3333) [reply .codecov.yml:6]: Copied this from the BoostReal repo, not sure what exclusion means in that context but thought it meant that we are not getting the coverage result of those, keeping only the coverage of the library itself.

File: .travis.yml (hunk 2 above)
Comment (aTom3333) [reply .travis.yml:58]: It is an artifact left from my importation of BoostReal files, I didn't check thoroughly to remove what doesn't makes sense here

**Review by sdavtaker — APPROVED** [review]

**Pull request merged**